File: JUNE-2025/29-06-2025/p1498.ex
defmodule Solution do
  @large 1_000_000_007

  @spec num_subseq(nums :: [integer], target :: integer) :: integer
  def num_subseq(nums, target) do
    Enum.sort(nums)
    |> Enum.split_while(&(&1 <= target / 2))
    |> then(fn {left, right} ->
      List.foldr(left, {0, 1, right, target}, &do_fold/2)
    end)
    |> elem(0)
  end

  defp do_fold(lo, {ans, combi, [hi | right], target}) when lo + hi <= target do
    do_fold(lo, {ans, rem(combi * 2, @large), right, target})
  end
  defp do_fold(lo, {ans, combi, right, target}) do
    {rem(ans + combi, @large), rem(combi * 2, @large), right, target}
  end
end

File: JUNE-2025/29-06-2025/p1498.dart
class Solution {
    int numSubseq(List<int> nums, int target) {
        nums.sort();
        int n = nums.length;
        int l = 0, r = n - 1;
        int res = 0;
        int mod = 1000000007;

        List<int> pows = new List.filled(n, 0);
        for (int i = 0 ; i < n ; i++)
            pows[i] = i == 0 ? 1 : pows[i - 1] * 2 % mod;

        while (l <= r) 
            if (nums[l] + nums[r] > target) 
                r--;
            else 
                res = (res + pows[r - l++]) % mod;
        return res;

    }
}